### PR TITLE
refactor: watch client api

### PR DESCRIPTION
**Pull request by LingKa28**
Base on #16 

Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)
Refactor watch client api to improve easy of use of the watch clinet feature.

* what changes does this pull request make?
  * refactor watch clinet api & test sample

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)
No.

### CODE REVIEW

File: client/lock.go [hunk 1]
```diff
@@ -238,33 +239,48 @@ func (c *lockClient) waitDelete(pfx string, myRev int64) {
...
+		// err = watcher.Send(&xlineapi.WatchRequest{
+		// 	RequestUnion: &xlineapi.WatchRequest_CreateRequest{
+		// 		CreateRequest: &xlineapi.WatchCreateRequest{
+		// 			Key: lastKey,
+		// 		},
+		// 	},
+		// })
+		// if err != nil {
+		// 	logger.Warn("Watch lease ID fail.", zap.Error(err))
+		// }
```
Comment (bsbds) [inline client/lock.go:254]: Please remove the comments

File: client/lock.go [hunk 2]
```diff
@@ -238,33 +239,48 @@ func (c *lockClient) waitDelete(pfx string, myRev int64) {
...
+		// 	for _, e := range watchRes.Events {
+		// 		if e.Type == xlineapi.Event_DELETE {
+		// 			f = true
+		// 			break
+		// 		}
+		// 	}
+		// 	if f {
+		// 		break
+		// 	}
+		// }
```
Comment (bsbds) [inline client/lock.go:283]: Same as above

File: client/lock.go [hunk 3]
```diff
@@ -238,33 +239,48 @@ func (c *lockClient) waitDelete(pfx string, myRev int64) {
...
+		// 	RequestUnion: &xlineapi.WatchRequest_CreateRequest{
+		// 		CreateRequest: &xlineapi.WatchCreateRequest{
+		// 			Key: lastKey,
+		// 		},
+		// 	},
+		// })
+		// if err != nil {
+		// 	logger.Warn("Watch lease ID fail.", zap.Error(err))
+		// }
+		for r := range watcher {
```
Comment (iGxnon) [inline client/lock.go:255]: Please remove them if we dont need

File: client/lock.go (hunk 2 above)
Comment (iGxnon) [inline client/lock.go:283]: Same as above

File: client/watch.go [hunk 4]
```diff
@@ -1,27 +1,107 @@
...
+						},
+					},
+				}
+				err := wc.Send(r)
+				if err != nil {
+					panic("watch cancel fail")
+				}
+				res, err := wc.Recv()
+				if err != nil {
+					panic("watch cancel fail")
```
Comment (bsbds) [inline client/watch.go:87]: please don't panic on error, use log or return an error instead.

File: client/lock.go [hunk 5]
```diff
@@ -238,33 +239,48 @@ func (c *lockClient) waitDelete(pfx string, myRev int64) {
...
 			f := false
-			watchRes, err := watcher.Recv()
-			if err != nil {
-				logger.Warn("Watch lease ID fail.", zap.Error(err))
-			}
-			for _, e := range watchRes.Events {
+			for _, e := range r.Events {
 				if e.Type == xlineapi.Event_DELETE {
 					f = true
```
Comment (iGxnon) [inline client/lock.go:259]: Will the assignment to f will be reordered?

File: client/watch.go [hunk 6]
```diff
@@ -1,27 +1,107 @@
...
+				}
+				wch <- (*WatchResponse)(res)
+				close(wch)
+				return
+			default:
+				res, err := wc.Recv()
+				if err != nil {
+					c.logger.Error("watch fail", zap.Error(err))
+				}
```
Comment (bsbds) [inline client/watch.go:96]: If the error not nil, should we return immediately?

**Review by themanforfree — APPROVED** [review]

File: client/lock.go (hunk 3 above)
Comment (LingKa28) [reply client/lock.go:255]: Done.

File: client/lock.go (hunk 2 above)
Comment (LingKa28) [reply client/lock.go:283]: Done.

File: client/lock.go (hunk 5 above)
Comment (LingKa28) [reply client/lock.go:259]: This is not a problem to be solved in this pr, the change is just to make the ci work, we can discuss that in #19

File: client/watch.go (hunk 6 above)
Comment (LingKa28) [reply client/watch.go:96]: I don't think it's necessary. It's maybe just a random failure, doesn't affect subsequent functions.

File: client/watch.go (hunk 4 above)
Comment (LingKa28) [reply client/watch.go:87]: I've modified as you suggested.

**Review by bsbds — APPROVED** [review]

**Review by bsbds — APPROVED** [review]

**Review by iGxnon — APPROVED** [review]

**Pull request merged**